File: Piece of cake/508. Most Frequent Subtree Sum/main.cpp
#include <bits/stdc++.h>
using namespace std;
// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right)
        : val(x), left(left), right(right) {}
};

class Solution {
   public:
    vector<int> findFrequentTreeSum(TreeNode *root) {
        unordered_map<int, int> cnt;
        vector<int> ans;
        int max_cnt = -1;
        function<int(TreeNode *)> dfs = [&](TreeNode *root) -> int {
            int sum = root->val;
            if (root->left) sum += dfs(root->left);
            if (root->right) sum += dfs(root->right);
            max_cnt = max(max_cnt, ++cnt[sum]);
            return sum;
        };
        dfs(root);
        for (auto [k, v] : cnt)
            if (v == max_cnt) ans.push_back(k);
        return ans;
    }
};

int main() { return 0; }

File: Piece of cake/508. Most Frequent Subtree Sum/main.rs
#![allow(dead_code)]
struct Solution;
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}
use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;
impl Solution {
    pub fn find_frequent_tree_sum(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        fn dfs(
            node: &Option<Rc<RefCell<TreeNode>>>,
            cnt: &mut HashMap<i32, i32>,
            max: &mut i32,
        ) -> i32 {
            if let Some(node) = node {
                let mut sum = node.borrow().val;
                sum += dfs(&node.borrow().left, cnt, max);
                sum += dfs(&node.borrow().right, cnt, max);
                let entry = cnt.entry(sum).or_default();
                *entry += 1;
                *max = *max.max(entry);
                sum
            } else {
                0
            }
        }

        let mut cnt = HashMap::new();
        let mut max = 0;
        dfs(&root, &mut cnt, &mut max);
        cnt.into_iter()
            .filter(|&(_, v)| v == max)
            .map(|(k, _)| k)
            .collect()
    }
}

fn main() {}
